File: solutions/sonnet-1/sonnet/ruby/calc.rb
class Parser
  def initialize(str)
    @s = str
    @pos = 0
    @len = str.length
  end

  def parse
    result = parse_expr
    result
  end

  private

  def skip_spaces
    @pos += 1 while @pos < @len && @s[@pos] == ' '
  end

  def peek
    skip_spaces
    @pos < @len ? @s[@pos] : nil
  end

  def parse_expr
    value = parse_term
    loop do
      c = peek
      if c == '+'
        @pos += 1
        value += parse_term
      elsif c == '-'
        @pos += 1
        value -= parse_term
      else
        break
      end
    end
    value
  end

  def parse_term
    value = parse_factor
    loop do
      c = peek
      if c == '*'
        @pos += 1
        value *= parse_factor
      elsif c == '/'
        @pos += 1
        divisor = parse_factor
        value = value.abs.div(divisor.abs) * ((value < 0) == (divisor < 0) ? 1 : -1)
      else
        break
      end
    end
    value
  end

  def parse_factor
    c = peek
    if c == '('
      @pos += 1
      value = parse_expr
      skip_spaces
      @pos += 1 if peek == ')'
      value
    else
      parse_number
    end
  end

  def parse_number
    skip_spaces
    start = @pos
    @pos += 1 while @pos < @len && @s[@pos] =~ /[0-9]/
    @s[start...@pos].to_i
  end
end

STDIN.each_line do |line|
  line = line.chomp
  next if line.strip.empty?
  puts Parser.new(line).parse
end

File: solutions/sonnet-1/sonnet/perl/calc.pl
use strict;
use warnings;

my ($str, $pos);

sub skip_ws {
    while ($pos < length($str) && substr($str, $pos, 1) eq ' ') {
        $pos++;
    }
}

sub parse_expr {
    my $val = parse_term();
    skip_ws();
    while ($pos < length($str) && (substr($str, $pos, 1) eq '+' || substr($str, $pos, 1) eq '-')) {
        my $op = substr($str, $pos, 1);
        $pos++;
        my $rhs = parse_term();
        if ($op eq '+') {
            $val += $rhs;
        } else {
            $val -= $rhs;
        }
        skip_ws();
    }
    return $val;
}

sub parse_term {
    my $val = parse_factor();
    skip_ws();
    while ($pos < length($str) && (substr($str, $pos, 1) eq '*' || substr($str, $pos, 1) eq '/')) {
        my $op = substr($str, $pos, 1);
        $pos++;
        my $rhs = parse_factor();
        if ($op eq '*') {
            $val *= $rhs;
        } else {
            my $q = $val / $rhs;
            $q = $q >= 0 ? int($q) : -int(-$q);
            $val = $q;
        }
        skip_ws();
    }
    return $val;
}

sub parse_factor {
    skip_ws();
    if (substr($str, $pos, 1) eq '(') {
        $pos++;
        my $val = parse_expr();
        skip_ws();
        if ($pos < length($str) && substr($str, $pos, 1) eq ')') {
            $pos++;
        }
        return $val;
    } else {
        skip_ws();
        my $start = $pos;
        while ($pos < length($str) && substr($str, $pos, 1) =~ /[0-9]/) {
            $pos++;
        }
        my $num = substr($str, $start, $pos - $start);
        return $num + 0;
    }
}

while (my $line = <STDIN>) {
    chomp $line;
    next if $line =~ /^\s*$/;
    $str = $line;
    $pos = 0;
    my $result = parse_expr();
    print "$result\n";
}
